### ISSUE TITLE
Autocomplete dropdown shows inconsistent results in `DataSearch` component

### ISSUE THREAD

**Issue by tankchintan** [appbaseio/reactivesearch#1148]
**Affected Projects**
React Web

**Library Version:**
`3.0.0-alpha.3`

**Describe the bug**
I use `customQuery` in a `DataSearch` component. When I type a query in the search box I see 2 issues,
1. The query generated by autocomplete (the one that shows instance results as a drop down list form the search box) is the stock query that is fired as if there was no `customQuery` being specified. This results in a disparity in results shown in the connected `ReactiveList` because I want `cross_fields` while searching, which is not present in stock query for data search component.
2. Even though there are results for a query in the autocomplete they are not being rendered. I found this out when I was wanted to customize what info was shown in the autocomplete using `parseSuggestion`. Here I printed the `suggestion` on console & I also monitor the query & response from server. I do not see the results as I would expect. Note that this does not happen all the time. 

**To Reproduce**
1. Basically specify a `customQuery` on `DataSearch` & see the query fired when you are typing in the search box versus when you finally hit the enter which then renders the `ReactiveList`
2. Again you need to monitor the `parseSuggestions` method & see what suggestions are being printed on the consiole versus what are actually shown in the autocomplete section.

**Expected behavior**
1. Use the same query as `customQuery` specified
2. Render all the suggestions/results form the query.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Brave/Chrome

**Additional context**

My `DataSearch` component invocation,
```
<DataSearch
  componentId="nameReactor"
  placeholder="Search for Name, Address, Phone, Email"
  dataField={['full_name', 'address', 'primary_email', 'primary_cell_phone']}
  fieldWeights={[10, 8, 6, 4]}
  searchInputId="universal-search"
  filterLabel="Search"
  iconPosition="right"
  autosuggest={true}
  fuzziness="AUTO"
  showClear={true}
  customQuery={
    function(value, props) {
      if (!value || !props) {
        return;
      }

      let fields = [];
      let currentDataField = props.dataField || [];
      let currentFieldWeights = props.fieldWeights || [];
      currentDataField.forEach(function(currentField, index) {
        let currentFieldWeight = currentField;
        if (index < currentFieldWeights.length) {
          currentFieldWeight += "^" + currentFieldWeights[index];
        }
        fields.push(currentFieldWeight);
      });

      const multiMatchTypes = ["best_fields", "cross_fields", "phrase_prefix"];
      const fuzzinessCapableTypes = ["best_fields"];

      let multiMatchDefinitions = [];
      multiMatchTypes.forEach(function(currentMultiMatchType) {

      let currentDefinition = {
            "query": value,
            "fields": fields,
            "type": currentMultiMatchType,
            "operator": "or"
      };

      if (props.fuzziness) {

        if (fuzzinessCapableTypes.includes(currentMultiMatchType)) {
            currentDefinition["fuzziness"] = props.fuzziness;
        }
      }

      multiMatchDefinitions.push({
                  "multi_match": currentDefinition
                })});

      return {
        "query": {
          "bool": {
              "should": multiMatchDefinitions,
              "minimum_should_match": "1"
          }          
        }
      }
    }
  }
parseSuggestion={
  function(suggestion) {

    if (!suggestion || suggestion === undefined) {
      return;
    }

    console.log("suggestion:", suggestion)

    return {
      label: (
          <div>
              {suggestion.source.first_name} {suggestion.source.middle_name} {suggestion.source.last_name}
               &nbsp;| {suggestion.source.address_line_1}
              <span style={{ color: 'dodgerblue', marginLeft: 5 }}>
                  {suggestion.source.zone}
              </span>
          </div>
      ),
      value: suggestion.value,
      source: suggestion.source  // for onValueSelected to work with parseSuggestion
    }
  }
}
renderError={error => (
    <div>
      Something went wrong with DataSearch
      <br />
      Error details
      <br />
      {error}
    </div>
  )}
  renderNoSuggestion={() => (
          <div>
              No suggestions found
          </div>
      )
  }
/>
```

I also keep on seeing the below error when I load the page,
```
index.js:1375 Warning: Failed prop type: Invalid prop `show` of type `boolean` supplied to `PoweredBy`, expected `number`.
    in PoweredBy (created by ReactiveList)
    in ReactiveList
    in Unknown (created by Context.Consumer)
    in Connect(Component) (created by ForwardRef)
    in ForwardRef (at App.js:348)
    in div (at App.js:346)
    in div (at App.js:227)
    in div
    in Unknown (created by Styled(Component))
    in Styled(Component) (created by URLParamsProvider)
    in URLParamsProvider
    in Unknown (created by Context.Consumer)
    in Connect(Component) (created by ForwardRef)
    in ForwardRef (created by ReactiveBase)
    in Provider (created by ReactiveBase)
    in ThemeProvider (created by ReactiveBase)
    in ReactiveBase (at App.js:78)
    in div (at App.js:77)
    in App (at src/index.js:7)
```

**Comment by tankchintan** [comment]
So seems like the dropdown issue is happening when the fields in the `dataField` are all derived fields. i.e. the content of those fields are due to use of `copy_to`

And per my understanding even though the ES sends back results there is some filtering going on in the reactiveSearch side which doesn't see the query clauses in the fields present in `dataField`, this is because technically there is no stored value for it, and I would keep it that way, since the only purpose of that fields is improve in search perf.

**Comment by bietkul** [comment]
It seems like a duplicate of #1002.

**Comment by tankchintan** [comment]
@bietkul thanks for responding.

I think it is similar to #1002 in the sense that the proposed solution might be the same. But I think the cause is different, in this case it might be genuinely an issue since the derived fields (or `copy_to`'ed target fields) by default will never have any stored content in it.

So looking at the proposed solutions,
1. Use label key for the title or define dataFields in a way that it won't replicate the source.
I don't think I understand this. I have the following in my `parseSuggestions`,
```
parseSuggestion={
                  function(suggestion) {
                    if (!suggestion || suggestion === undefined) {
                      return;
                    }
                    return {
                      label: (
                          <div>
                              {suggestion.source.first_name} {suggestion.source.middle_name} {suggestion.source.last_name}                             
                          </div>
                      ),
                      value: suggestion.value,
                      source: suggestion.source  // for onValueSelected to work with parseSuggestion
                    }
                  }
                }
```
Do you propose to do something else here?

2. If the first method doesn't work then you can use custom render function to render the suggestions.
I did this as well, no luck. This is the updated `DataSearch`,
```
<DataSearch
  componentId="nameReactor"
  placeholder="Search for Name, Address, Phone, Email"
  dataField={['full_name', 'address.search', 'emails.*gram', 'phones.autocomplete']}
  fieldWeights={[10, 8, 6, 5]}
  searchInputId="universal-search"
  filterLabel="Search"
  iconPosition="right"
  autosuggest={true}
  fuzziness="AUTO"
  showClear={true}
  debounce={150}
  customQuery={
      ...
      return {
        //generated query
    }
  }
  render={({
      loading,
      error,
      data,
      value,
      downshiftProps: {
          isOpen,
          getItemProps,
      }
  }) => {
      if(loading) {
          return <div>Fetching Suggestions.</div>
      }
      if(error) {
          return (
              <div>
                  Something went wrong! Error details {JSON.stringify(error)}
              </div>
          )
      }
      console.log(data, value)
      return (Boolean(value.length)) ? (
                  <div>
                      {data.slice(0, 5).map((suggestion, index) => (
                          <div
                              key={suggestion.value}
                              {...getItemProps({ item: suggestion })}
                          >
                              {suggestion.value}
                          </div>
                      ))}
                      {Boolean(value.length) && (
                          <div {...getItemProps({ item: { label: value, value: value }})}>
                              Show all results for "{value}"
                          </div>
                      )}
                  </div>
      ) : null;
  }}  
/>
```
This still does not return any results. The `data` array is empty inside `render`.

**Comment by bietkul** [comment]
The second solution will work for you now, use `rawData` instead of `data`.
Please check the docs of `render` prop [here](https://docs.appbase.io/docs/reactivesearch/v3/search/datasearch/).

We can have a new prop in the search components to disable this behavior, currently, we parse the suggestions which may lead to duplicates if someone customizes the `label` key.
For now, the only solution is to use custom renderer.

**Comment by bietkul** [comment]
Regarding the custom query issue( empty `data` key), can you create a codesandbox example.

**Comment by tankchintan** [comment]
@bietkul I am not sure I follow. I don't think I have any custom query issue. I was facing issue with `data` being empty. But that was because `data` contains all the filtered response, as opposed to `rawData`.

**Comment by bietkul** [comment]
Ok, it seems like you're getting hits in `rawData`.
We'll soon introduce a new prop to disable this parsing logic.

**Comment by ShahAnuj2610** [comment]
@tankchintan right now it is not feasible to include a `prop`, that would control whether suggestions should be parsed or not. Since suggestions are formed of the shape `{label, value, source}`, it is not possible to parse suggestions keeping the no of items same for multiple `dataFields`. If we have to implement this feature, we have to assume that `dataFields` will always be single.

**Comment by tankchintan** [comment]
Thanks for the note @ShahAnuj2610 
I guess my question is, is it possible to instead have a `prop` which determines whether or not to allow more suggestions than requested? i.e. if there exists a prop `maxNumSuggestions` (I dont think I found it), can it be instead converted to `minNumSuggestions`?

Or we could have a separate prop instead, `allowMoreSuggestions`?

Or we could just show the max suggestions, assuming the more relevant suggestions are at first, so just slice the suggestion array.

So if the hold up is around having same number of items, maybe the above methods cover that? 

Just a thought.

**Comment by priyanksha-monotype** [comment]
@tankchintan any update on how to show more than 10 results

**Comment by tankchintan** [comment]
Priyanka, unfortunately no updates. I was not able to do what I needed.

**Comment by ShahAnuj2610** [comment]
@priyanksha @tankchintan it is possible to show more than 10 suggestions in both `React` and `Vue`. You can use the `size` prop fixed by following two PRs:

- React: https://github.com/appbaseio/reactivesearch/pull/1323
- Vue: https://github.com/appbaseio/reactivesearch/pull/1344

**Comment by jyash97** [comment]
Hey @tankchintan thanks for reporting the issue 💯 
Duplicate suggestions issue has been solved now. We have added a prop called `showDistinctSuggestions` which when set to `true` parses 1 suggestion per document. For more detail, you can refer the docs [here](https://docs.appbase.io/docs/reactivesearch/v3/search/datasearch/).

**Comment by arpithparikh** [comment]
showDistinctSuggestions is not returning 1 suggestion per document. Could you please fix it? At least on react side of it.

**Comment by LarsFronius** [comment]
Hmm, are we sure that showDistinctSuggestions is fixed @bietkul ?

Sample Config for me:
```typescript
<DataSearch
    componentId="search"
    dataField={[
        'Title',
        'Description',
        'SKU',
        'EAN'
    ]}
    fieldWeights={[
        2,
        1,
        3,
        3
    ]}
    fuzziness={2}
    queryFormat={'and'}
    react={{
        and: ['category', 'brand']
    }}
    showDistinctSuggestions={true}
})}
/>
```

Still shows multiple suggestions for the same doc.

**Comment by bietkul** [comment]
@LarsFronius Can you share a replicable codesandbox link?

**Comment by bietkul** [comment]
@LarsFronius I was able to replicate the issue. We have fixed it in `v3.24.1`.

**Comment by LarsFronius** [comment]
@bietkul just to give an update - sorry for not providing a codesandbox link, but I can confirm the fix is working. Thanks for the great work here!
